include/nlohmann/json.hpp

```diff
@@ -3555,15 +3804,128 @@ class basic_json
         return value(ptr, string_t(default_value));
     }
 
+    /*!
+    @brief return the binary subtype
+
+    Returns the numerical subtype of the JSON value, if the JSON value is of
+    type "binary", and it has a subtype.  If it does not have a subtype (or the
+    object is not of type binary) this function will return size_t(-1) as a
+    sentinel value.
+
+    @return the numerical subtype of the binary JSON value
+
+    @complexity Constant.
+
+    @exceptionsafety No-throw guarantee: this member function never throws
+    exceptions.
+
+    @sa @ref set_subtype() -- sets the binary subtype
+    @sa @ref clear_subtype() -- clears the binary subtype
+    @sa @ref has_subtype() -- returns whether or not the binary value has a
+    subtype
+
+    @since version 3.8.0
+    */
+    std::size_t get_subtype() const noexcept
+    {
+        if (is_binary() and m_value.binary->has_subtype)
+        {
+            return m_value.binary->subtype;
+        }
+
+        return std::size_t(-1);
+    }
+
+    /*!
+    @brief sets the binary subtype
+
+    Sets the binary subtype of the JSON value, also flags a binary JSON value as
+    having a subtype, which has implications for serialization to msgpack (will
+    prefer ext file formats over bin).  If the JSON value is not a binary value,
+    this function does nothing.
+
+    @complexity Constant.
+
+    @exceptionsafety No-throw guarantee: this member function never throws
+    exceptions.
+
+    @sa @ref get_subtype() -- return the binary subtype
+    @sa @ref clear_subtype() -- clears the binary subtype
+    @sa @ref has_subtype() -- returns whether or not the binary value has a
+    subtype
+
+    @since version 3.8.0
+    */
+
+    void set_subtype(std::uint8_t subtype) noexcept
+    {
+        if (is_binary())
+        {
+            m_value.binary->has_subtype = true;
+            m_value.binary->subtype = subtype;
+        }
+    }
+
+    /*!
+    @brief clears the binary subtype
+
+    Clears the binary subtype of the JSON value, also flags a binary JSON value
+    as not having a subtype, which has implications for serialization to msgpack
+    (will prefer bin file formats over ext).  If the JSON value is not a binary
+    value, this function does nothing.
+
+    @complexity Constant.
+
+    @exceptionsafety No-throw guarantee: this member function never throws
+    exceptions.
+
+    @sa @ref get_subtype() -- return the binary subtype
+    @sa @ref set_subtype() -- sets the binary subtype
+    @sa @ref has_subtype() -- returns whether or not the binary value has a
+    subtype
+
+    @since version 3.8.0
+    */
+    void clear_subtype() noexcept
+    {
+        if (is_binary())
+        {
+            m_value.binary->has_subtype = false;
```

Comment:
I would additionally set the subtype back to its initial value `0` to remove some WTF moments in debugging.

Reply:
This shouldn't actually matter, as the only way (through this interface) to set the subtype is via the `set_subtype` command.  But I've resolved this in d767c1c7